include/envoy/http/filter.h

```diff
@@ -190,10 +190,17 @@ class StreamDecoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual void addDecodedData(Buffer::Instance& data, bool streaming_filter) PURE;
 
-  /**
-   * Adds decoded trailers. This will overwrite any existing trailers should any already exist.
-   */
-  virtual void addDecodedTrailers(HeaderMapPtr&& trailers) PURE;
+   /**
+    * Adds decoded trailers. May only be called in decodeData when end_stream is set to true or in
+    * decodeTrailers. If called in any other context, std::logic_error will be thrown.
```

Comment:
Can you call this under decodeTrailers?    isn't decodeTrailers the wrong state (not LastDataFrame) and would have request_trailers_ non-null (exception either way)?

Reply:
Also I'm not sure if we're consistent about this but can you add a javadoc style @throws here and below?

Reply:
Yeah you're right - I was misunderstanding when `request_trailers_` was set in the normal flow. The test failures were due to exactly this.

I'll add in the `@throws` as well.